### ISSUE TITLE
Inserting data via `nodetool refresh --load-and-stream` fails silently when a column was dropped and recreated

### ISSUE THREAD

**Issue by racevedoo** [scylladb/scylladb#18264]
When inserting new data via `nodetool refresh --load-and-stream` after a column was deleted and recreated with the same name, Scylla ignores the existence of the column.

<details>
<summary>Bash script to reproduce</summary>

Assuming a scylla instance is running locally with docker (via `docker run --name scylla -d scylladb/scylla:5.2.16`)
```bash
#!/bin/bash

set -ueox pipefail

docker exec -it scylla cqlsh -e "create keyspace if not exists ks with replication = {'class': 'SimpleStrategy', 'replication_factor': 1};"
docker exec -it scylla cqlsh -e "create table if not exists ks.t (a int primary key, b int);"
docker exec -it scylla cqlsh -e "insert into ks.t (a, b) values (1, 2);"
docker exec -it scylla cqlsh -e "select * from ks.t;"
docker exec -it scylla nodetool flush ks t
docker exec -it scylla bash -c "rm -rf /sstables; mkdir -p /sstables; cp -r /var/lib/scylla/data/ks/t-*/me-*-big-* /sstables/"
docker exec -it scylla bash -c 'for i in $(ls /sstables/me-*-big-Data.db); do sstabledump $i; done'
docker exec -it scylla cqlsh -e "alter table ks.t drop b;"
docker exec -it scylla cqlsh -e "select * from ks.t;"
docker exec -it scylla cqlsh -e "alter table ks.t add b int;"
docker exec -it scylla cqlsh -e "select * from ks.t;"
docker exec -it scylla bash -c 'for i in $(ls /sstables/me-*-big-*); do cp $i /var/lib/scylla/data/ks/t-*/upload/; done'
docker exec -it scylla nodetool refresh ks t --load-and-stream
docker exec -it scylla cqlsh -e "select * from ks.t;"
```
</details>

The script output is as follows:
```
+ docker exec -it scylla cqlsh -e 'create keyspace if not exists ks with replication = {'\''class'\'': '\''SimpleStrategy'\'', '\''replication_factor'\'': 1};'
+ docker exec -it scylla cqlsh -e 'create table if not exists ks.t (a int primary key, b int);'
+ docker exec -it scylla cqlsh -e 'insert into ks.t (a, b) values (1, 2);'
+ docker exec -it scylla cqlsh -e 'select * from ks.t;'

 a | b
---+---
 1 | 2

(1 rows)
+ docker exec -it scylla nodetool flush ks t
+ docker exec -it scylla bash -c 'rm -rf /sstables; mkdir -p /sstables; cp -r /var/lib/scylla/data/ks/t-*/me-*-big-* /sstables/'
+ docker exec -it scylla bash -c 'for i in $(ls /sstables/me-*-big-Data.db); do sstabledump $i; done'
[
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2024-04-16T18:35:40.933278Z" },
        "cells" : [
          { "name" : "b", "value" : 2 }
        ]
      }
    ]
  }
]+ docker exec -it scylla cqlsh -e 'alter table ks.t drop b;'
+ docker exec -it scylla cqlsh -e 'select * from ks.t;'

 a
---
 1

(1 rows)
+ docker exec -it scylla cqlsh -e 'alter table ks.t add b int;'
+ docker exec -it scylla cqlsh -e 'select * from ks.t;'

 a | b
---+------
 1 | null

(1 rows)
+ docker exec -it scylla bash -c 'for i in $(ls /sstables/me-*-big-*); do cp $i /var/lib/scylla/data/ks/t-*/upload/; done'
+ docker exec -it scylla nodetool refresh ks t --load-and-stream
+ docker exec -it scylla cqlsh -e 'select * from ks.t;'

 a | b
---+------
 1 | null

(1 rows)
```

IMO, the last select statement should return the output below, since nodetool refresh ran successfully.
```
 a | b
---+---
 1 | 2

(1 rows)
```

Is this the expected behavior? I couldn't find anything in the docs

*Installation details*
Scylla version (or git commit hash): 5.2.16
Cluster size: 1
OS (RHEL/CentOS/Ubuntu/AWS AMI): Ubuntu (via docker image)

**Comment by denesb** [comment]
Yes, the new column `b` is not the same as the dropped one, even if they have the same name. When restoring the sstables, all values for the old dropped `b` column is filtered out, because ScyllaDB detects that the data from the sstable was written before the column was dropped.

**Comment by denesb** [comment]
@tzach @annastuchlik this is possibly something we need to document. When dropping a column, the column will be dropped from restored backup as well, even if alter recreated with the same name.

**Comment by racevedoo** [comment]
Maybe there should be a config flag that blocks recreating columns that were dropped with the same name? This seems to cause weird behaviors all over the place. 
Just ran into a scenario where the table, materialized view and cache return different results for the same column (it was dropped and recreated with the same name in the past).

Something like https://github.com/scylladb/scylladb/blob/master/cql3/statements/alter_table_statement.cc#L156-L159 should be enough

**Comment by denesb** [comment]
Makes sense. Indeed the results are counter-intuitive and looks like we may have subtle bugs around this in the code.

**Comment by mykaul** [comment]
> Maybe there should be a config flag that blocks recreating columns that were dropped with the same name? This seems to cause weird behaviors all over the place.

People (not machines) make mistakes. Are we going to punish them forever for their mistakes? How long and where shall we keep those deleted column names?

**Comment by racevedoo** [comment]
> People (not machines) make mistakes. Are we going to punish them forever for their mistakes? How long and where shall we keep those deleted column names?

Agreed, that's why I'm suggesting a **opt-in** config flag that blocks re-creating a column with the same name as the one that was dropped, since that can trigger weird behaviors. People who do not set this flag will get the current behavior.

Regarding where to keep deleted column names, I think ScyllaDB already keeps them in the `system_schema.dropped_columns` table. Am I correct?

**Comment by racevedoo** [comment]
Hacked [this commit](https://github.com/scylladb/scylladb/commit/49894f0d0f0ff0db476083e499ab6b71ff0c5e2f) (it needs a lot of refinement) but it seems to work. If it makes sense, I can refine it and open a PR for further discussion.

**Comment by denesb** [comment]
> > Maybe there should be a config flag that blocks recreating columns that were dropped with the same name? This seems to cause weird behaviors all over the place.
> 
> People (not machines) make mistakes. Are we going to punish them forever for their mistakes? How long and where shall we keep those deleted column names?

I think we keep the names of deleted columns names forever (while the table exists). These are needed to be able to read old sstables, written before the column was dropped.

**Comment by denesb** [comment]
> While going through the code, I came up with [this commit](https://github.com/scylladb/scylladb/commit/49894f0d0f0ff0db476083e499ab6b71ff0c5e2f) (it needs a lot of refinement) but it seems to work. The unit test makes the option pretty clear.
> 
> If it makes sense, I can refine it and open a PR for further discussion.

We need input from product here (@tzach). What do we do: add a safety option, which can be constraining, or document our way out of this?

**Comment by tzach** [comment]
Lets start with documention.

**Comment by mykaul** [comment]
@tzach - did we document this?

**Comment by mykaul** [comment]
re-ping @tzach

**Comment by mykaul** [comment]
re-ping @tzach
